[로또 미션] 남해윤 미션 제출합니다

haeyoon1 opened the pull request:
(수정 전)
controller 
domain
view 
지난 미션때 패키지를 나누는것에 익숙하지 않아 세개의 패키지를 나누어 코드를 작성하는 것에 집중하였습니다. 
최대한 기능을 나눠 코드를 작성했지만 아직 미숙한 것 같습니다..  알려주신 리뷰 바탕으로 열심히 수정하겠습니다. 잘부탁드립니다!


(수정 후)
view 에서 Lotto를 생성하는 부분에대해 리뷰남겨주신 것을 보고, 로또를 생성하는 LottoFactory 클래스를 domain 패키지에 생성했습니다.
또한 Statistic 클래스에서 if문이 많은 부분에 관해서는 다음 코드리뷰때 알려주신다고 하셨는데, enum을 사용해서 수정하는것이 맞을까요? 맞다면 최종 merge하기 전에 수정하겠습니다!
test코드를 제외하고 모두 수정하고, 남겨주신 코멘트에 질문 및 답변 남겼는데 확인해주시면 감사하겠습니다! 잘부탁드립니다.

haeyoon1 committed:
로또 구입 금액을 입력하면 구입 금액에 해당하는 로또를 발급해야 한다.
로또 1장의 가격은 1000원이다.

haeyoon1 committed:
로또 당첨 번호를 받아 일치한 번호 수에 따라 당첨 결과를 보여준다.

haeyoon1 committed:
로또 당첨 번호를 받아 일치한 번호 수에 따라 당첨 결과를 보여준다.

haeyoon1 committed:
2등을 위한 보너스볼을 추첨한다.
당첨 통계에 2등을 추가한다.
2등 당첨 조건은 당첨 번호 5개 일치 + 보너스 볼 일치다.

haeyoon1 committed:
현재 로또 생성기는 자동 생성 기능만 제공한다. 사용자가 수동으로 추첨 번호를 입력할 수 있도록 해야 한다.
입력한 금액, 자동 생성 숫자, 수동 생성 번호를 입력하도록 해야 한다.

haeyoon1 committed:
테스트 코드 작성

haeyoon1 committed:
코드 수정
- 로또를 생성하는 LottoFactory 클래스 추가

haeyoon1 committed:
테스트 코드 작성

70825 commented on src/main/java/LottoApplication.java:
```diff
@@ -0,0 +1,9 @@
+public class LottoApplication {
+
+    public static void main(String[] args) {
+        final LottoController controller = new LottoController();
```
P2) 다른 곳에서는 `final`을 사용하지 않는데, 여기는 `final`을 사용한 이유가 있을까요??

70825 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,57 @@
+
+    public static final Scanner scanner = new Scanner(System.in);
+
+    public static int readPrice(){
```
P2) `InputView`와 `OutputView`에서는 `static` 메서드들을 사용헀는데 `static`을 사용한 이유가 있을까요??

haeyoon1 replied:
> P2) `InputView`와 `OutputView`에서는 `static` 메서드들을 사용헀는데 `static`을 사용한 이유가 있을까요??

객체를 변경하지 않고, 단순 입력을 받거나 출력을 하는 역할만 하기 때문에 정적메서드를 사용했습니다!

70825 replied:
👍👍

70825 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,57 @@
+        return scanner.nextInt();
+    }
+
+    public static List<Lotto> handLottosNumber(int numberOfHandLottos){ //수동 로또 번호 입력 메서드
```
P1) `InputView`에서 `List<Lotto>`를 생성하고 있네요. 저는 `View`는 화면 입출력에만 관여를 해야한다고 생각하는데, 어떻게 생각하시나요?? `LottoWinningNumber(...)`도 마찬가지 입니다.

70825 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,57 @@
+        return lottoList;
+    }
+
+    public static List<Integer> LottoWinningNumber (int numberOfLottos) { //로또 당첨 번호 입력 메서드
```
P1) 메서드 네이밍은 일반적으로 카멜케이스라서 `InputView.java`랑 `Statistics.java`쪽 메서드를 수정해주세요

70825 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,57 @@
+        List<Integer> winningNumbers = new ArrayList<>();
+        for (String number: splitNumbers){
+            winningNumbers.add(Integer.parseInt(number));
+        }
```
P2) 전반적인 코드를 확인해보니 어디는 stream을 사용하고, 어디는 for문을 사용하고 있더라구요. 해윤님이 stream, for를 사용하는 기준이 따로 있나요??

70825 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,42 @@
+            if (prize == Prize.FIVE_MATCHES_BONUS){
+                System.out.println("5개 일치, 보너스 볼 일치(" + prize.getPrizeAmount() + "원)- " + matchCounts.get(prize) + "개");
+            }
+            //else 쓰는게 맞는 것 같지만 조건때문에 if문으로 변경
+            if (prize != Prize.FIVE_MATCHES_BONUS) {
+                System.out.println(prize.getMatchingCount() + "개 일치 (" + prize.getPrizeAmount() + "원)- " + matchCounts.get(prize) + "개");
```
P3) 오 여기에 중요한 주석이 있네요. else 대신 if를 쓰면 어떤 장점이 있을까요??

haeyoon1 replied:
> P3) 오 여기에 중요한 주석이 있네요. else 대신 if를 쓰면 어떤 장점이 있을까요??

해당 경우에는 else를 사용하는 것이 더 효율적인 것 같은데 **기존프로그래밍 요구사항**에
**_else 예약어를 쓰지 않는다.
else 예약어를 쓰지 말라고 하니 switch/case로 구현하는 경우가 있는데 switch/case도 허용하지 않는다.
힌트: if문에서 값을 반환하는 방식으로 구현하면 else 예약어를 사용하지 않아도 된다._**
라고 되어있어서 if문으로 작성했습니다. 위의 코드에서도 if문을 사용하는 것이 좋은건가요?

70825 replied:
음 지금 코드에서는 간단하기 때문에 if, else 둘 다 큰 차이는 없지만, 위 요구사항에 else문 대신 if문으로 쓰라는 이유는 맨 처음에 나온 if문을 계속 기억할 필요가 없기 때문에 일반적으로는 else 대신 if문을 쓰는 편이에요.

```
if (조건 1) {
    ....
} else {
    ....
}
```

이거보다는 `else` 대신 `if (조건1의 반대)`라는게 `if(조건 1)`을 꼬아서 생각할 필요가 사라지니까요.

70825 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,59 @@
+
+public class LottoController {
+
+    private static final int LOTTO_PRICE = 1000;
```
매직 넘버 상수화 👍👍

70825 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,59 @@
+        int numberOfAutoLottos = numberOfLottos - numberOfHandLottos;
+
+        //수동 로또 번호 입력
+        List<Lotto> handLottos = InputView.handLottosNumber(numberOfHandLottos);
```
P2) `Lottos`라는 일급컬렉션을 만들어 두셨는데, `List<Lotto>`를 사용하는 이유가 있나요??

haeyoon1 replied:
> P2) `Lottos`라는 일급컬렉션을 만들어 두셨는데, `List<Lotto>`를 사용하는 이유가 있나요??

Lotto와 관련된 비즈니스 로직까지 추가해서 Lottos 클래스를 생성했습니다. 사실 명확한 이유는 잘 모르는 상태로 이전 자동차경주 미션때 많은 분들이 Car, Cars 클래스를 생성했던 것 처럼 이번에도 코드를 작성했는데.. 코드의 응집도를 향상시킬 수 있도록 사용한 것 같습니다.
혹시 제가 코드에서 일급컬렉션을 불필요하게 사용한 것같다면 알려주세요!

70825 replied:
불필요하게 사용하기보다는 `List<Lotto>` 대신 `Lottos`로 수정해서 오히려 `Lottos`를 더 많이 사용할 수 있어보여요
지금 로또 미션에서는 `Lottos`를 사용하는 이유가 `List<Lotto>` 감싸기 말고는 보이지 않네요.. 자동차 경주 미션으로 예를 들자면 `Cars`로 `Cars.java` 내부에서 `Cars`만 할 수 있는 로직을 만들 수 있어서 말씀하신 코드의 응집도와 관련이 있어요. (ex. 자동차들중에 우승차 구하기, 자동차들 움직이게 하기, ...)
다음 미션을 살펴봤는데 다음 미션에서는 일급 컬렉션의 필요성을 느낄 수 있을 것 같아보이네요

70825 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,59 @@
+        //수동 로또 번호 입력
+        List<Lotto> handLottos = InputView.handLottosNumber(numberOfHandLottos);
+
+        //자동 로또 번호 생성
```
주석 있어서 보기 너무 편하네요 👍

70825 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,59 @@
+        for (Lotto lotto : finalLottos) {
+            int matchedCount = lotto.countMatches(winningNumbers);
+            boolean isBonusMatch = lotto.contains(bonusNumber);
+            statistics.updateStatistics(matchedCount, isBonusMatch);  // 통계 업데이트
```
P2) 여기서는 `statistics.updateStatistics(...)`라는 메서드인데, 통계라는 글자가 중복되어 있네요. 더 좋은 네이밍으로 수정 가능해보여요

70825 commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,57 @@
+public class Lotto{
+
+    private static final int LOTTO_NUMBER_COUNT = 6;
+    private List<Integer> lottoNumbers;
```
```suggestion
    private static final int LOTTO_NUMBER_COUNT = 6;

    private List<Integer> lottoNumbers;
```

P3) 보통 이렇게 상수랑 변수는 줄바꿈으로 구분하는 편이에요

70825 commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,57 @@
+        List<Integer> selectedNumbers = numbers.subList(0,6);
+
+        Collections.sort(selectedNumbers);
+        // ArrayList의 contains() 메소드는 어디 사용되는지 모르겠습니다.
```
P3) 이거는 아래에 `public boolean contains(...)` 메서드에서 `return lottoNumbers.contains(bonusNumber);`로 잘 활용하셨는데, 어느 부분이 궁금하신건가요?

70825 commented on src/main/java/domain/Lottos.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class Lottos {
```
P3) 이거는 학습 차원에서 시도해보는걸 추천드려요

`LottoController`에서 `Lottos autolottos = new Lottos(numberOfAutoLottos);`라는 내용 코드가 있는데요. 이 부분을 정적 팩토리 메서드를 만들어서 적용해볼까요?? ([참고할 글](https://sun-22.tistory.com/84))

70825 commented on src/main/java/domain/Lottos.java:
```diff
@@ -0,0 +1,29 @@
+        List<Lotto> lottoList = new ArrayList<>();
+        for (int i=0; i<numberOfAutoLottos; i++){
+            lottoList.add(new Lotto());
+        }
```
P3) 이거도 학습 차원으로 Stream API를 사용할 수 있을 것 같아요

70825 commented on src/main/java/domain/NumberGenerater.java:
```diff
@@ -0,0 +1,6 @@
+package domain;
+
+public interface NumberGenerater {
+
+    int generate();
+}
```
P2) 이거는 인터페이스를 만들어둔 이유가 있나요??

70825 commented on src/main/java/domain/Statistics.java:
```diff
@@ -0,0 +1,46 @@
+
+public class Statistics {
+
+    private final Map<Prize, Integer> matchCounts = new HashMap<>();
```
P2) 학습 차원에서 `HashMap`말고 `EnumMap`을 사용해보는건 어떨까요?? ([참고 블로그](https://yeonyeon.tistory.com/195))

70825 commented on src/main/java/domain/Statistics.java:
```diff
@@ -0,0 +1,46 @@
+
+    //(일반 번호)맞은 개수 업데이트하는 메서드
+    public void updateStatistics(int matchCount, boolean isBonusMatch){
+        if (matchCount == 6){
+            CountUp(Prize.SIX_MATCHES);
+        } else if (matchCount == 5 && isBonusMatch){
+            CountUp(Prize.FIVE_MATCHES_BONUS);
+        } else if (matchCount == 5){
+            CountUp(Prize.FIVE_MATCHES);
+        } else if (matchCount == 4){
+            CountUp(Prize.FOUR_MATCHES);
+        } else if (matchCount == 3){
+            CountUp(Prize.THREE_MATCHES);
+        }
```
P3) 이 부분은 개선이 필요한데, 다음 코드 리뷰 시간으로 넘길게요. 일단은 함수형 인터페이스에 대해 공부하면 좋아보여요

70825 commented on src/test/java/domain/LottoTest.java:
```diff
@@ -0,0 +1,55 @@
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class LottoTest {
```
```java
@SuppressWarnings("NonAsciiCharacters") // 한글과 같은 아스키 코드가 아닌 문자에 밑줄이 쳐지지 않음
@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class) // UnderScore(_)가 공백으로 표시
class LottoTest {
```

P3) 한글명으로 테스트 이름을 작성하면 인텔리제이에 밑줄이 보이기도 하고, 테스트 실행하면 언더바가 보이는데 이걸 없애주는 방법이에요

70825 commented on src/test/java/domain/LottoTest.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @Test
+    void 자동_로또_생성_테스트(){
```
P1) 하나의 테스트 코드에 3개를 검증하고 있네요. 이 부분은 테스트 하나에 목적 하나만 확인할 수 있도록 분리하는게 좋아보입니다.

- 로또 개수는 6개이다
- 로또 번호의 범위는 1~45 사이이다
- 로또 번호는 중복이 없어야한다

자동 로또 생성 테스트라는 큰 틀을 사용하고 싶으면 `@Nested`를 사용하는 방법도 있어요

```java
@Nested
class 자동_로또_생성_테스트 {

    @Test
    void 로또_개수는_6개야_한다 () {
        ...
    }
```

이건 테스트 코드를 전체적으로 업데이트하는게 좋아보여요

70825 replied:
P2) 학습 차원으로 given - when - then 스타일을 사용해보는건 어떨까요?

70825 commented on src/test/java/domain/LottoTest.java:
```diff
@@ -0,0 +1,55 @@
+        assertEquals(6, uniqueCount, "로또 번호는 중복이 없어야합니다.");
+    }
+
+    @Test
+    void 수동_로또_생성_테스트(){
+        List<Integer> generatedNumbers = handLotto.getLottoNumbers();
+
+        // 테스트할 내용이 없는 것 같습니다..
+    }
```
P3) 테스트할 내용이 없으면 삭제하는게 좋아보여요. 나중에 필요하면 그때 추가하면 되니까요

70825 reviewed (changes requested):
안녕하세요 해윤님~ 미션 코드 잘 봤습니다. 그런데 PR 제목 바꾸셔야 할 것 같아요

변수명도 좋고, 컨트롤러에 주석이 있어서 읽기 편했네요 👍👍

제가 이전 리뷰 코멘트는 시간날 때 코멘트를 올릴 예정이라 리뷰 반영 후 다시 코드 리뷰를 받고 싶으시면 맨 아래 새로운 코멘트로 리뷰해달라고 이야기해주세요

코멘트에 우선순위를 달아두었어요. P1은 가급적 고쳐야하는 내용, P2는 의견 나눠보고 고치고 싶다하면 고치는 내용, P3은 참고하는 내용입니다. P2, P3은 저도 궁금한 내용이라 같이 의견 나누어봐요

---

### P1) 인텔리제이에 적용할 내용들

1) 파일 개행 설정

어느 파일은 개행이 되어 있고, 어느 곳은 개행이 되어 있지 않아서 인텔리제이에 설정해두면 [좋은 글](https://hstory0208.tistory.com/entry/IntelliJ-%ED%8C%8C%EC%9D%BC-%EB%81%9D%EC%97%90-%EC%9E%90%EB%8F%99%EC%9C%BC%EB%A1%9C-%EA%B0%9C%ED%96%89-%EC%B6%94%EA%B0%80%ED%95%98%EA%B8%B0-No-newline-at-end-of-file) 공유합니다

2) import 와일드 카드 제거

Import 문으로는 필요한 내용만 불러야 한다고 생각해서 여기도 [블로그 글](https://hianna.tistory.com/835) 공유할게요

3) 코드 포맷팅하기

코딩할 때 `a = 1`로 쓰는게 보통 케이스지만, 가끔가다 `a=1`로 작성한 것처럼 의도치 않게 규칙적이지 않은 컨벤션이 나오게 되는데요. 이걸 제거하려면 [이 글](https://asj8000.tistory.com/entry/Intellij-%EC%BD%94%EB%93%9C-%EC%9E%90%EB%8F%99-%EC%A0%95%EB%A0%AC-%EB%8B%A8%EC%B6%95%ED%82%A4-%EB%B0%8F-%EC%BB%A4%EC%8A%A4%ED%84%B0%EB%A7%88%EC%9D%B4%EC%A7%95)을 보면 좋아보여요.

이 기능은 리뷰 받기 직전에 한 번만 해주시고 커밋 날리면 됩니다!

고생하셨어요~!!

haeyoon1 committed:
Merge branch 'main' into haeyoon1

haeyoon1 committed:
코드 수정
- 로또를 생성하는 LottoFactory 클래스 추가

70825 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,52 @@
+import domain.Lotto;
+import domain.LottoFactory;
+
+import java.util.*;
```
P1) 와일드카드가 아직 있네요!

70825 commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,59 @@
+
+    //로또 번호 수동 생성 생성자
+    public Lotto(List<Integer> numbers) {
+        if (numbers.size() != LOTTO_NUMBER_COUNT) {
+            throw new IllegalStateException("로또 번호는 6개여야 합니다.");
+        }
```
P2) 이 부분은 로또를 생성하는 것에 집중된 코드가 아닌, 로또를 생성하기 전에 검증하는 로직에 가까운데, 이 부분만 메서드를 하나 더 만들어서 분리하는건 어떨까요??

70825 commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,59 @@
+
+        List<Integer> numbers = new ArrayList<>();
+
+        for (int i = 1; i <= 45; i++) {
```
P2) `LottoController.java`의 `LOTTO_PRICE`처럼 1, 45를 상수화 시킬 수 있어보여요. 아래에 `number.subList(0, 6)`도 마찬가지입니다. 각각 어떤 네이밍이 좋을까요??

70825 commented on src/main/java/domain/LottoFactory.java:
```diff
@@ -0,0 +1,29 @@
+public class LottoFactory {
+
+    //수동 로또 생성 메서드
+    public static List<Lotto> createManualLottos(List<List<Integer>> manualNumbers) {
+        return manualNumbers.stream()
+                .map(Lotto::new)
+                .collect(Collectors.toList());
+    }
```
P1) 파라미터 타입이 `List<List<Integer>>`면 나중에 유지보수할 때 `List<List<Integer>>`의 의미를 파악하기 힘들어보여요. 이 부분은 `InputView.java`에서 한 번에 처리하려고 해서 발생한 문제로 보이는데 `InputView.java`의 `handLottosNumber`를 어떻게 분리할 수 있을까요??

haeyoon1 replied:
새로 ManualLottoNumbers 클래스를 만들고 수정은 했으나, 해당 부분이 멘토님의 의도대로 수정된 것인지 잘 모르겠습니다..

70825 replied:
아녀 저는 `List<List<Integer>>` 이거 리스트를 두 번 감싼 것을 하나로 줄이길 원했어요
`InputView`의 `handLottosNumber`에서 for문으로 돌려 한 번에 모든 로또를 얻기보다 외부에서 반복문을 사용해서 `InputView` 입장에서는 수동 로또 입력 받는 메서드가 수동 번호 개수만큼 호출되는 것이죠.

70825 commented on src/test/java/domain/LottoTest.java:
```diff
@@ -0,0 +1,59 @@
+    private Lotto autoLotto;
+    private Lotto handLotto;
+
+    @BeforeEach
+    void setUp() {
+        autoLotto = new Lotto();
+        handLotto = new Lotto(Arrays.asList(1, 2, 3, 4, 5, 6));
+    }
```
P3) `@BeforeEach`를 사용하면 어떤 장점이 있고, 어떤 단점이 예상되나요?

70825 commented on src/main/java/domain/Lottos.java:
```diff
@@ -0,0 +1,22 @@
+        return LottoFactory.createAutoLottos(numberOfAutoLottos);
+    }
+
+    public List<Lotto> getLottos() {
+        return lottos;
+    }
```
P2) [불변 객체](https://mangkyu.tistory.com/131)도 공부해보면 좋아보여요
코드 자체는 `Collections.unmodifiableList(lotto);`로 수정하는 것 뿐이라 현재는 왜 수정해야하는지 와닿지 않는 상황이겠지만, 위 글을 통해 어떤 효과를 얻을 수 있는지 알아두면 좋아보여요!

저는 `4. 부수 효과(Side Effect)를 피해 오류가능성을 최소화할 수 있다.`, `5. 다른 사람이 작성한 함수 + 내가 작성한 함수를 다른 사람들이 예측가능하며 안전하게 사용할 수 있다.`, `불변성(Immutability)이 보장된 코드 리딩` 여기에 대해 중요하다고 느껴졌습니다.

이거는 아마 팀프로젝트나 서로 협업하는 일이 있을 때 필요성이 느껴질 확률이 높아보이네요

70825 reviewed (changes requested):
고생하셨어요 해윤님 🚀🚀

반영 다 하시면 맨 아래에 스크롤 내리시면 새로운 댓글 입력하는 곳이 있는데, 댓글에 @70825처럼 저 멘션 후에 지금처럼 궁금한 부분이나 리뷰 반영 다 했다는 내용을 남기시면 됩니다~

> view 에서 Lotto를 생성하는 부분에대해 리뷰 남겨주신 것을 보고, 로또를 생성하는 LottoFactory 클래스를 domain 패키지에 생성했습니다.

오호 팩토리 클래스를 따로 생성한 이유가 있을까요?? 저는 `Lotto.java`, `Lottos.java`에 두어도 충분한 내용으로 보여서요.

> enum을 사용해서 수정하는것이 맞을까요?

오 네 맞아요. 첫 리뷰에 코멘트가 많기도 하고, 함수형 인터페이스는 저는 처음 공부할 때 받아들이는게 어려웠어요.

그래서 첫 리뷰에서 전반적인 부분은 거의 끝내두고 함수형 인터페이스쪽을 반영하게 만들 계획이었네요.

저도 `enum`으로 수정할 방향을 생각했는데, 이미 방법을 아시는 것 같으니 수정해볼까요?? 새로운 `enum`을 만드셔도 되고, `Prize`쪽에 만들어도 됩니다.

haeyoon1 committed:
refactor: 와일드카드 제거 및 코드 정리

- validateLottoNumbers 메서드 분리
- 와일드카드 제거
- 매직 넘버 상수화

haeyoon1 committed:
refactor(LottoFactory, InputView): ManualLottoNumbers 클래스를 추가하고 관련 메서드 수정

ManualLottoNumbers 클래스를 새로 생성하고, LottoFactory의 createManualLottos 메서드와 InputView의 handLottosNumber 메서드를 수정했습니다.

해당 피드백에 대해 헷갈림이 있어, 검토가 필요합니다.

haeyoon1 committed:
refactor: Prize enum을 사용하여 맞춘 개수에 따른 상금 결정 로직 개선

- Statistics 클래스의 matchCountUpdate 메서드에서 Prize enum의 fromMatchCount 메서드를 호출하도록 수정.
- IllegalArgumentException을 사용하여 유효하지 않은 맞춘 개수에 대한 예외 처리 추가.

haeyoon1 commented:
기간은 지났지만 아직 merge되지 않은 것 같아서 피드백 주신 내용 반영해서 최종 수정했습니다! 멘토님 덕분에 많이 배웠습니다. 완벽하게 수정하지는 못했지만 좋은 피드백 너무 감사했습니다:)

70825 reviewed (approved):
해윤님 고생하셨어요 ~ 
우테코 7기 미션도 같이 병행중이시군요 ㄷㄷ
좋은 결과 있길 바라겠습니다 🚀🚀
